frontend/lib/resolution/call-graph.cpp

```diff
@@ -0,0 +1,206 @@
+#include "chpl/resolution/resolution-types.h"
+#include "chpl/uast/Function.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/TypeDecl.h"
+
+namespace chpl {
+namespace resolution {
+
+
+using namespace uast;
+using namespace types;
+
+struct CalledFnCollector {
+  using RV = ResolvedVisitor<CalledFnCollector>;
+
+  // input
+  Context* context;
+  const AstNode* symbol = nullptr; // Module* or Function*
+  const ResolvedFunction* resolvedFunction = nullptr; // set if not module
+  int depth = 0;
+
+  // output
+  CalledFnsSet& called;
+
+  CalledFnCollector(Context* context,
+                    const AstNode* symbol,
+                    const ResolvedFunction* resolvedFunction,
+                    int depth,
+                    CalledFnsSet& called)
+    : context(context),
+      symbol(symbol),
+      resolvedFunction(resolvedFunction),
+      depth(depth),
+      called(called)
+  {
+  }
+
+  // helper to run the visitor on something
+  void process();
+  // collect calls from the ResolvedExpression
+  void collect(const ResolvedExpression* re);
+
+  bool enter(const Module* mod, RV& rv) {
+    return mod == symbol; // only proceed if it's the module requested
+  }
+  void exit(const Module* mod, RV& rv) {
+  }
+
+  bool enter(const TypeDecl* ast, RV& rv) {
+    return false;
+  }
+  void exit(const TypeDecl* ast, RV& rv) {
+  }
+
+  bool enter(const Function* fn, RV& rv) {
+    return fn == symbol; // only proceed if it's the function requested
+  }
+  void exit(const Function* fn, RV& rv) {
+  }
+
+  bool enter(const AstNode* ast, RV& rv) {
+    if (auto re = rv.byPostorder().byAstOrNull(ast)) {
+      collect(re);
+    }
+    return true;
+  }
+  void exit(const AstNode* ast, RV& rv) {
+  }
+};
+
+
+void CalledFnCollector::process() {
+  if (resolvedFunction) {
+    CHPL_ASSERT(symbol && symbol->isFunction());
+    // we are handling a function, so visit based on the ResolvedFunction.
+    chpl::resolution::ResolutionContext rcval(context);
+    const ResolutionResultByPostorderID& byPostorder =
+      resolvedFunction->resolutionById();
+    ResolvedVisitor<CalledFnCollector> rv(&rcval, symbol, *this, byPostorder);
+    symbol->traverse(rv);
+  } else {
+    CHPL_ASSERT(symbol && symbol->isModule());
+    chpl::resolution::ResolutionContext rcval(context);
+    const ResolutionResultByPostorderID& byPostorder =
+      resolveModule(context, symbol->id());
+    ResolvedVisitor<CalledFnCollector> rv(&rcval, symbol, *this, byPostorder);
+    symbol->traverse(rv);
+  }
+}
+
+void CalledFnCollector::collect(const ResolvedExpression* re) {
+  const PoiScope* poiScope = re->poiScope();
+
+  // consider the return-intent overloads
+  for (const auto& candidate : re->mostSpecific()) {
+    if (const TypedFnSignature* sig = candidate.fn()) {
+      if (sig->untyped()->idIsFunction()) {
+        chpl::resolution::ResolutionContext rcval(context);
+        const ResolvedFunction* fn = resolveFunction(&rcval, sig, poiScope);
```

Comment:
Am I right to assume that this doesn't consider nested functions? It is my understanding that the resolutioncontext is supposed to keep track of the "call stack" of functions that were used to reach a given fn. But here, we're constructing a new resolution context from scratch, which means that called nested functions will lose their nested function calling context.

Reply:
Yes. I developed this code before David's recent work was merged. We can leave adjusting it to handle nested functions correctly as future work. It might require some changes more on the nested functions infrastructure side.

Reply:
I've added this to a Future Work section of the PR message.